src/inmanta/agent/agent.py

```diff
@@ -137,52 +135,184 @@ def __init__(self, scheduler: "ResourceScheduler", gid: uuid.UUID, reason: str)
         super().__init__(scheduler, dummy_id, gid, reason)
 
 
-class ResourceAction(ResourceActionBase):
-    def __init__(self, scheduler: "ResourceScheduler", resource: Resource, gid: uuid.UUID, reason: str) -> None:
+@dataclass(frozen=True)
+class ResourceInstallSpec:
+    """
+    This class encapsulates the requirements for a specific resource type for a specific model version.
+
+    :ivar resource_type: fully qualified name for this resource type e.g. std::File
+    :ivar model_version: the version of the model to use
+    :ivar pip_config: the pip config to use during requirements installation
+    :ivar requirements: python packages that must be installed prior to executing the module sources
+    :ivar sources: list of ModuleSource containing the code for deployment of this resource
+
+    """
+
+    resource_type: str
+    model_version: int
+    pip_config: PipConfig
+    requirements: Sequence[str]
+    sources: Sequence["ModuleSource"]
+
+
+class ResourceDetails:
+    """
+    In memory representation of the desired state of a resource
+    """
+
+    id: Id
+    rid: ResourceIdStr
+    rvid: ResourceVersionIdStr
+    env_id: uuid.UUID
+    model_version: int
+    requires: Sequence[Id]
+    attributes: dict[str, object]
+
+    def __init__(self, resource_dict: JsonType) -> None:
+        self.attributes = resource_dict["attributes"]
+        self.attributes["id"] = resource_dict["id"]
```

Comment:
two questions here:
1. why don't we keep the resource_dict itself? I would expect it to be required?
2. why modify the attributes, seems risky?

Reply:
@sanderr please correct me if I'm wrong there,

> 1. why don't we keep the resource_dict itself? I would expect it to be required?

Do you have something specific in mind for which it is required ? As far as I can tell current behavior is in line with the original implementation where we use:
- the raw JsonType resource's `resource_type` attribute to get the required code (now in `get_code`, previously in `ensure_code`
- the raw JsonType resource's `attributes` attribute during resource deserialization (now happening later in the flow, in the executor)

> 2. why modify the attributes, seems risky?

The end goal is that we need an `id` field to be present in the dict passed to `Resource.deserialize(<dict>)`. Would it be better if add the `id` field to `resource_dict["attributes"]` prior to the `ResourceDetails` constructor call ? That's what the old implementation did but I don't think it is any better.

An alternative would be to to something like `Resource.deserialize({**attributes, 'id': rvid})` ?

> seems risky 

Why is it risky I'm curious ?

Reply:
I think both 1 and 2 have remained essientially unchanged versus master. Can you confirm, @Hugo-Inmanta?

Reply:
yes